---
title: "Figure 3"
output: html_notebook
---

# Dependencies
```{r}
library(Seurat)
library(dplyr)
library(viridis)
library(pheatmap)
library(mgcv)
```

# Load the data
```{r}
a549_egf <- readRDS("../data/A549_EGF.rds")
a549_tnf <- readRDS("../data/A549_TNF.rds")
a549_tgfb1 <- readRDS("../data/A549_TGFB1.rds")

du145_egf <- readRDS("../data/DU145_EGF.rds")
du145_tnf <- readRDS("../data/DU145_TNF.rds")
du145_tgfb1 <- readRDS("../data/DU145_TGFB1.rds")

mcf7_egf <- readRDS("../data/MCF7_EGF.rds")
mcf7_tnf <- readRDS("../data/MCF7_TNF.rds")
mcf7_tgfb1 <- readRDS("../data/MCF7_TGFB1.rds")

ovca420_egf <- readRDS("../data/OVCA420_EGF.rds")
ovca420_tnf <- readRDS("../data/OVCA420_TNF.rds")
ovca420_tgfb1 <- readRDS("../data/OVCA420_TGFB1.rds")
```


# Fig 3a
## Heatmap of canonical TFs
Code is super verbose, but the goal here is simply to make a binarized heatmap of which
datasets specific TFs are differentially expressed in

```{r}
a549_tgfb1_gam <- read.csv("../output/pseudotime/a549_tgfb1_dge.csv", row.names = 1)
a549_egf_gam <- read.csv("../output/pseudotime/a549_egf_dge.csv", row.names=1)
a549_tnf_gam <- read.csv("../output/pseudotime/a549_tnf_dge.csv", row.names=1)

du145_tgfb1_gam <- read.csv("../output/pseudotime/du145_tgfb1_dge.csv", row.names=1)
du145_egf_gam <- read.csv("../output/pseudotime/du145_egf_dge.csv", row.names=1)
du145_tnf_gam <- read.csv("../output/pseudotime/du145_tnf_dge.csv", row.names=1)

mcf7_tgfb1_gam <- read.csv("../output/pseudotime/mcf7_tgfb1_dge.csv", row.names=1)
mcf7_egf_gam <- read.csv("../output/pseudotime/mcf7_egf_dge.csv", row.names=1)
mcf7_tnf_gam <- read.csv("../output/pseudotime/mcf7_tnf_dge.csv", row.names=1)

ovca420_tgfb1_gam <- read.csv("../output/pseudotime/ovca420_tgfb1_dge.csv", row.names=1)
ovca420_egf_gam <- read.csv("../output/pseudotime/ovca420_egf_dge.csv", row.names=1)
ovca420_tnf_gam <- read.csv("../output/pseudotime/ovca420_tnf_dge.csv", row.names=1)
```

```{r}
a549_tgfb1_sig <- rownames(a549_tgfb1_gam)[which(a549_tgfb1_gam$fdr <= 0.05)]
a549_egf_sig <- rownames(a549_egf_gam)[which(a549_egf_gam$fdr <= 0.05)]
a549_tnf_sig <- rownames(a549_tnf_gam)[which(a549_tnf_gam$fdr <= 0.05)]

du145_tgfb1_sig <- rownames(du145_tgfb1_gam)[which(du145_tgfb1_gam$fdr <= 0.05)]
du145_egf_sig <- rownames(du145_egf_gam)[which(du145_egf_gam$fdr <= 0.05)]
du145_tnf_sig <- rownames(du145_tnf_gam)[which(du145_tnf_gam$fdr <= 0.05)]

mcf7_tgfb1_sig <- rownames(mcf7_tgfb1_gam)[which(mcf7_tgfb1_gam$fdr <= 0.05)]
mcf7_egf_sig <- rownames(mcf7_egf_gam)[which(mcf7_egf_gam$fdr <= 0.05)]
mcf7_tnf_sig <- rownames(mcf7_tnf_gam)[which(mcf7_tnf_gam$fdr <= 0.05)]

ovca420_tgfb1_sig <- rownames(ovca420_tgfb1_gam)[which(ovca420_tgfb1_gam$fdr <= 0.05)]
ovca420_egf_sig <- rownames(ovca420_egf_gam)[which(ovca420_egf_gam$fdr <= 0.05)]
ovca420_tnf_sig <- rownames(ovca420_tnf_gam)[which(ovca420_tnf_gam$fdr <= 0.05)]
```

```{r}
#Make a function so we can tweak the number of genes
getVarGenes <- function(seurat_object, n){
  var_table <- seurat_object[["RNA"]]@meta.features
  var_genes <- rownames(var_table[order(desc(var_table$vst.variance.standardized)),])[1:n]
}
```

```{r}
a549_tgfb1_sig <- a549_tgfb1_sig[which(a549_tgfb1_sig %in% getVarGenes(a549_tgfb1, 2000))]
a549_egf_sig <- a549_egf_sig[which(a549_egf_sig %in% getVarGenes(a549_egf, 2000))]
a549_tnf_sig <- a549_tnf_sig[which(a549_tnf_sig %in% getVarGenes(a549_tnf, 2000))]

du145_tgfb1_sig <- du145_tgfb1_sig[which(du145_tgfb1_sig %in% getVarGenes(du145_tgfb1, 2000))]
du145_egf_sig <- du145_egf_sig[which(du145_egf_sig %in% getVarGenes(du145_egf, 2000))]
du145_tnf_sig <- du145_tnf_sig[which(du145_tnf_sig %in% getVarGenes(du145_tnf, 2000))]

mcf7_tgfb1_sig <- mcf7_tgfb1_sig[which(mcf7_tgfb1_sig %in% getVarGenes(mcf7_tgfb1, 2000))]
mcf7_egf_sig <- mcf7_egf_sig[which(mcf7_egf_sig %in% getVarGenes(mcf7_egf, 2000))]
mcf7_tnf_sig <- mcf7_tnf_sig[which(mcf7_tnf_sig %in% getVarGenes(mcf7_tnf, 2000))]

ovca420_tgfb1_sig <- ovca420_tgfb1_sig[which(ovca420_tgfb1_sig %in% getVarGenes(ovca420_tgfb1, 2000))]
ovca420_egf_sig <- ovca420_egf_sig[which(ovca420_egf_sig %in% getVarGenes(ovca420_egf, 2000))]
ovca420_tnf_sig <- ovca420_tnf_sig[which(ovca420_tnf_sig %in% getVarGenes(ovca420_tnf, 2000))]
```

```{r}
deg_list <- data.frame(Gene = unique(c(a549_tgfb1_sig , a549_egf_sig, a549_tnf_sig,
                                       du145_tgfb1_sig, du145_egf_sig, du145_tnf_sig,
                                       mcf7_tgfb1_sig, mcf7_egf_sig, mcf7_tnf_sig,
                                       ovca420_tgfb1_sig, ovca420_egf_sig, ovca420_tnf_sig)))
  
deg_list$A549_TGFB1 <- deg_list$Gene %in% a549_tgfb1_sig
deg_list$A549_EGF <- deg_list$Gene %in% a549_egf_sig
deg_list$A549_TNF <- deg_list$Gene %in% a549_tnf_sig

deg_list$DU145_TGFB1 <- deg_list$Gene %in% du145_tgfb1_sig
deg_list$DU145_EGF <- deg_list$Gene %in% du145_egf_sig
deg_list$DU145_TNF <- deg_list$Gene %in% du145_tnf_sig

deg_list$MCF7_TGFB1 <- deg_list$Gene %in% mcf7_tgfb1_sig
deg_list$MCF7_EGF <- deg_list$Gene %in% mcf7_egf_sig
deg_list$MCF7_TNF <- deg_list$Gene %in% mcf7_tnf_sig

deg_list$OVCA420_TGFB1 <- deg_list$Gene %in% ovca420_tgfb1_sig
deg_list$OVCA420_EGF <- deg_list$Gene %in% ovca420_egf_sig
deg_list$OVCA420_TNF <- deg_list$Gene %in% ovca420_tnf_sig

#Convert TRUE/FALSE to 1/0
deg_list[,2:ncol(deg_list)] <- lapply(deg_list[,2:ncol(deg_list)], as.numeric)

rownames(deg_list) <- deg_list$Gene
deg_list <- deg_list[,2:ncol(deg_list)]
```

```{r}
emt_tfs <- c("SNAI1", "SNAI2", "ZEB1", "ZEB2", "TWIST1", "TWIST2")
emt_tf_mat <- deg_list[emt_tfs,]
rownames(emt_tf_mat) <- emt_tfs
emt_tf_mat[is.na(emt_tf_mat)] <- 0
```

```{r}
emt_tf_heatmap <- pheatmap(emt_tf_mat,
                          color = c("grey95", "red"),
                          border_color="black",
                          cluster_cols=F,
                          cluster_rows=F,
                          show_rownames=F,
                          show_colnames=F,
                          legend=F,
                          clustering_method="ward.D2",
                          file="../figure_panels/figure_3/EMTTFs_PseudoDGE_Heatmap.png",
                          width=2.25, height=1.3)
```

# Fig 3b
## Sum of TF counts
```{r}
a549_tgfb1_aucell <- read.csv("../output/SCENIC/a549_tgfb1_auc_mtx.csv",
                              row.names=1)
a549_egf_aucell <- read.csv("../output/SCENIC/a549_egf_auc_mtx.csv",
                              row.names=1)
a549_tnf_aucell <- read.csv("../output/SCENIC/a549_tnf_auc_mtx.csv",
                              row.names=1)

du145_tgfb1_aucell <- read.csv("../output/SCENIC/du145_tgfb1_auc_mtx.csv",
                              row.names=1)
du145_egf_aucell <- read.csv("../output/SCENIC/du145_egf_auc_mtx.csv",
                              row.names=1)
du145_tnf_aucell <- read.csv("../output/SCENIC/du145_tnf_auc_mtx.csv",
                              row.names=1)

mcf7_tgfb1_aucell <- read.csv("../output/SCENIC/mcf7_tgfb1_auc_mtx.csv",
                              row.names=1)
mcf7_egf_aucell <- read.csv("../output/SCENIC/mcf7_egf_auc_mtx.csv",
                              row.names=1)
mcf7_tnf_aucell <- read.csv("../output/SCENIC/mcf7_tnf_auc_mtx.csv",
                              row.names=1)

ovca420_tgfb1_aucell <- read.csv("../output/SCENIC/ovca420_tgfb1_auc_mtx.csv",
                              row.names=1)
ovca420_egf_aucell <- read.csv("../output/SCENIC/ovca420_egf_auc_mtx.csv",
                              row.names=1)
ovca420_tnf_aucell <- read.csv("../output/SCENIC/ovca420_tnf_auc_mtx.csv",
                              row.names=1)
```

```{r}
runGAM <- function(seurat_object, aucell){
  colnames(aucell) <- gsub("[...]", "", colnames(aucell))
  exp <- t(as.matrix(aucell))
  #Get rid of odd case of all zero values
  exp <- exp[rowSums(exp)!=0,]
  
  gam.pval <- apply(exp,1,function(z){
    d <- data.frame(exp=z, pseudotime=seurat_object$Pseudotime,
                    batch=seurat_object$Mix)
    tmp <- mgcv::gam(exp ~ s(pseudotime, k=4) + batch, 
                     data=d,
                     method="REML")
    p <- summary(tmp)[[8]] # where the p-value is stored
    p
    })
  gam.pval <- as.data.frame(gam.pval)
  colnames(gam.pval) <- "pval"
  gam.pval$Gene <- rownames(gam.pval)
  gam.pval$fdr <- p.adjust(gam.pval$pval, method='fdr')
  return(gam.pval)
}
```

Differential regulon activity over pseudotime
```{r}
a549_tgfb1_reg_test <- runGAM(a549_tgfb1, a549_tgfb1_aucell)
a549_egf_reg_test <- runGAM(a549_egf, a549_egf_aucell)
a549_tnf_reg_test <- runGAM(a549_tnf, a549_tnf_aucell)

du145_tgfb1_reg_test <- runGAM(du145_tgfb1, du145_tgfb1_aucell)
du145_egf_reg_test <- runGAM(du145_egf, du145_egf_aucell)
du145_tnf_reg_test <- runGAM(du145_tnf, du145_tnf_aucell)

mcf7_tgfb1_reg_test <- runGAM(mcf7_tgfb1, mcf7_tgfb1_aucell)
mcf7_egf_reg_test <- runGAM(mcf7_egf, mcf7_egf_aucell)
mcf7_tnf_reg_test <- runGAM(mcf7_tnf, mcf7_tnf_aucell)

ovca420_tgfb1_reg_test <- runGAM(ovca420_tgfb1, ovca420_tgfb1_aucell)
ovca420_egf_reg_test <- runGAM(ovca420_egf, ovca420_egf_aucell)
ovca420_tnf_reg_test <- runGAM(ovca420_tnf, ovca420_tnf_aucell)
```

Get genes differentially expressed over pseudotime
```{r}
a549_tgfb1_gam <- read.csv("../output/pseudotime/a549_tgfb1_dge.csv", row.names = 1)
a549_egf_gam <- read.csv("../output/pseudotime/a549_egf_dge.csv", row.names=1)
a549_tnf_gam <- read.csv("../output/pseudotime/a549_tnf_dge.csv", row.names=1)

du145_tgfb1_gam <- read.csv("../output/pseudotime/du145_tgfb1_dge.csv", row.names=1)
du145_egf_gam <- read.csv("../output/pseudotime/du145_egf_dge.csv", row.names=1)
du145_tnf_gam <- read.csv("../output/pseudotime/du145_tnf_dge.csv", row.names=1)

mcf7_tgfb1_gam <- read.csv("../output/pseudotime/mcf7_tgfb1_dge.csv", row.names=1)
mcf7_egf_gam <- read.csv("../output/pseudotime/mcf7_egf_dge.csv", row.names=1)
mcf7_tnf_gam <- read.csv("../output/pseudotime/mcf7_tnf_dge.csv", row.names=1)

ovca420_tgfb1_gam <- read.csv("../output/pseudotime/ovca420_tgfb1_dge.csv", row.names=1)
ovca420_egf_gam <- read.csv("../output/pseudotime/ovca420_egf_dge.csv", row.names=1)
ovca420_tnf_gam <- read.csv("../output/pseudotime/ovca420_tnf_dge.csv", row.names=1)
```

```{r}
a549_tgfb1_sig <- rownames(a549_tgfb1_gam)[which(a549_tgfb1_gam$fdr <= 0.05)]
a549_egf_sig <- rownames(a549_egf_gam)[which(a549_egf_gam$fdr <= 0.05)]
a549_tnf_sig <- rownames(a549_tnf_gam)[which(a549_tnf_gam$fdr <= 0.05)]

du145_tgfb1_sig <- rownames(du145_tgfb1_gam)[which(du145_tgfb1_gam$fdr <= 0.05)]
du145_egf_sig <- rownames(du145_egf_gam)[which(du145_egf_gam$fdr <= 0.05)]
du145_tnf_sig <- rownames(du145_tnf_gam)[which(du145_tnf_gam$fdr <= 0.05)]

mcf7_tgfb1_sig <- rownames(mcf7_tgfb1_gam)[which(mcf7_tgfb1_gam$fdr <= 0.05)]
mcf7_egf_sig <- rownames(mcf7_egf_gam)[which(mcf7_egf_gam$fdr <= 0.05)]
mcf7_tnf_sig <- rownames(mcf7_tnf_gam)[which(mcf7_tnf_gam$fdr <= 0.05)]

ovca420_tgfb1_sig <- rownames(ovca420_tgfb1_gam)[which(ovca420_tgfb1_gam$fdr <= 0.05)]
ovca420_egf_sig <- rownames(ovca420_egf_gam)[which(ovca420_egf_gam$fdr <= 0.05)]
ovca420_tnf_sig <- rownames(ovca420_tnf_gam)[which(ovca420_tnf_gam$fdr <= 0.05)]
```

Only those in the top 2k variable genes of each data set

```{r}
#Make a function so we can tweak the number of genes
getVarGenes <- function(seurat_object, n){
  var_table <- seurat_object[["RNA"]]@meta.features
  var_genes <- rownames(var_table[order(desc(var_table$vst.variance.standardized)),])[1:n]
}
```

```{r}
a549_tgfb1_sig <- a549_tgfb1_sig[which(a549_tgfb1_sig %in% getVarGenes(a549_tgfb1, 2000))]
a549_egf_sig <- a549_egf_sig[which(a549_egf_sig %in% getVarGenes(a549_egf, 2000))]
a549_tnf_sig <- a549_tnf_sig[which(a549_tnf_sig %in% getVarGenes(a549_tnf, 2000))]

du145_tgfb1_sig <- du145_tgfb1_sig[which(du145_tgfb1_sig %in% getVarGenes(du145_tgfb1, 2000))]
du145_egf_sig <- du145_egf_sig[which(du145_egf_sig %in% getVarGenes(du145_egf, 2000))]
du145_tnf_sig <- du145_tnf_sig[which(du145_tnf_sig %in% getVarGenes(du145_tnf, 2000))]

mcf7_tgfb1_sig <- mcf7_tgfb1_sig[which(mcf7_tgfb1_sig %in% getVarGenes(mcf7_tgfb1, 2000))]
mcf7_egf_sig <- mcf7_egf_sig[which(mcf7_egf_sig %in% getVarGenes(mcf7_egf, 2000))]
mcf7_tnf_sig <- mcf7_tnf_sig[which(mcf7_tnf_sig %in% getVarGenes(mcf7_tnf, 2000))]

ovca420_tgfb1_sig <- ovca420_tgfb1_sig[which(ovca420_tgfb1_sig %in% getVarGenes(ovca420_tgfb1, 2000))]
ovca420_egf_sig <- ovca420_egf_sig[which(ovca420_egf_sig %in% getVarGenes(ovca420_egf, 2000))]
ovca420_tnf_sig <- ovca420_tnf_sig[which(ovca420_tnf_sig %in% getVarGenes(ovca420_tnf, 2000))]
```

We'll filter the TF regulon differential expression results on FDR and ensure that the TF has to be differentially expressed
```{r}
a549_tgfb1_reg_sig <- filter(a549_tgfb1_reg_test, fdr < 0.05)$Gene
a549_egf_reg_sig <- filter(a549_egf_reg_test, fdr < 0.05)$Gene
a549_tnf_reg_sig <- filter(a549_tnf_reg_test, fdr < 0.05)$Gene

du145_tgfb1_reg_sig <- filter(du145_tgfb1_reg_test, fdr < 0.05)$Gene
du145_egf_reg_sig <- filter(du145_egf_reg_test, fdr < 0.05)$Gene
du145_tnf_reg_sig <- filter(du145_tnf_reg_test, fdr < 0.05)$Gene

mcf7_tgfb1_reg_sig <- filter(mcf7_tgfb1_reg_test, fdr < 0.05)$Gene
mcf7_egf_reg_sig <- filter(mcf7_egf_reg_test, fdr < 0.05)$Gene
mcf7_tnf_reg_sig <- filter(mcf7_tnf_reg_test, fdr < 0.05)$Gene

ovca420_tgfb1_reg_sig <- filter(ovca420_tgfb1_reg_test, fdr < 0.05)$Gene
ovca420_egf_reg_sig <- filter(ovca420_egf_reg_test, fdr < 0.05)$Gene
ovca420_tnf_reg_sig <- filter(ovca420_tnf_reg_test, fdr < 0.05)$Gene
```

```{r}
a549_tgfb1_reg_sig <- a549_tgfb1_reg_sig[which(a549_tgfb1_reg_sig %in% a549_tgfb1_sig)]
a549_egf_reg_sig <- a549_egf_reg_sig[which(a549_egf_reg_sig %in% a549_egf_sig)]
a549_tnf_reg_sig <- a549_tnf_reg_sig[which(a549_tnf_reg_sig %in% a549_tnf_sig)]

du145_tgfb1_reg_sig <- du145_tgfb1_reg_sig[which(du145_tgfb1_reg_sig %in% du145_tgfb1_sig)]
du145_egf_reg_sig <- du145_egf_reg_sig[which(du145_egf_reg_sig %in% du145_egf_sig)]
du145_tnf_reg_sig <- du145_tnf_reg_sig[which(du145_tnf_reg_sig %in% du145_tnf_sig)]

mcf7_tgfb1_reg_sig <- mcf7_tgfb1_reg_sig[which(mcf7_tgfb1_reg_sig %in% mcf7_tgfb1_sig)]
mcf7_egf_reg_sig <- mcf7_egf_reg_sig[which(mcf7_egf_reg_sig %in% mcf7_egf_sig)]
mcf7_tnf_reg_sig <- mcf7_tnf_reg_sig[which(mcf7_tnf_reg_sig %in% mcf7_tnf_sig)]

ovca420_tgfb1_reg_sig <- ovca420_tgfb1_reg_sig[which(ovca420_tgfb1_reg_sig %in% ovca420_tgfb1_sig)]
ovca420_egf_reg_sig <- ovca420_egf_reg_sig[which(ovca420_egf_reg_sig %in% ovca420_egf_sig)]
ovca420_tnf_reg_sig <- ovca420_tnf_reg_sig[which(ovca420_tnf_reg_sig %in% ovca420_tnf_sig)]
```

Count how many data sets each regulon is differentially expressed in
```{r}
deg_list <- data.frame(Gene = unique(c(a549_tgfb1_reg_sig , a549_egf_reg_sig, a549_tnf_reg_sig,
                                       du145_tgfb1_reg_sig, du145_egf_reg_sig, du145_tnf_reg_sig,
                                       mcf7_tgfb1_reg_sig, mcf7_egf_reg_sig, mcf7_tnf_reg_sig,
                                       ovca420_tgfb1_reg_sig, ovca420_egf_reg_sig, ovca420_tnf_reg_sig)))
  
deg_list$A549_TGFB1 <- deg_list$Gene %in% a549_tgfb1_reg_sig
deg_list$A549_EGF <- deg_list$Gene %in% a549_egf_reg_sig
deg_list$A549_TNF <- deg_list$Gene %in% a549_tnf_reg_sig

deg_list$DU145_TGFB1 <- deg_list$Gene %in% du145_tgfb1_reg_sig
deg_list$DU145_EGF <- deg_list$Gene %in% du145_egf_reg_sig
deg_list$DU145_TNF <- deg_list$Gene %in% du145_tnf_reg_sig

deg_list$MCF7_TGFB1 <- deg_list$Gene %in% mcf7_tgfb1_reg_sig
deg_list$MCF7_EGF <- deg_list$Gene %in% mcf7_egf_reg_sig
deg_list$MCF7_TNF <- deg_list$Gene %in% mcf7_tnf_reg_sig

deg_list$OVCA420_TGFB1 <- deg_list$Gene %in% ovca420_tgfb1_reg_sig
deg_list$OVCA420_EGF <- deg_list$Gene %in% ovca420_egf_reg_sig
deg_list$OVCA420_TNF <- deg_list$Gene %in% ovca420_tnf_reg_sig

#Convert TRUE/FALSE to 1/0
deg_list[,2:ncol(deg_list)] <- lapply(deg_list[,2:ncol(deg_list)], as.numeric)
```

```{r}
deg_list_counts <- as.data.frame(rowSums(deg_list[2:ncol(deg_list)]))
colnames(deg_list_counts) <- "Count"
deg_list_counts$Gene <- deg_list$Gene
deg_list_counts <- arrange(deg_list_counts, desc(Count))
deg_list_counts$Index <- 1:nrow(deg_list_counts)
```

```{r}
count_plot <- ggplot(deg_list_counts, aes(x=Index, y=Count)) +
  geom_point(size=0.25, color="black") +
  xlab("") + ylab("Significant Counts") +
  scale_y_continuous(breaks=c(0, 3, 6, 9, 12)) +
  theme_classic() +
  theme(axis.text=element_text(size=5, color="black"),
        axis.title=element_blank(),
        axis.line = element_line(size=0.25),
        axis.ticks = element_line(size=0.25))
ggsave(count_plot, filename = "../figure_panels/figure_3/TF_Regulon_DiffExp_Counts.png",
       dpi=600, width=3, height=2.5, units="cm")
```

```{r}
deg_list_counts
```


# Fig 3c
## Heatmap of TF coefficient
```{r}
linearGAM <- function(seurat_object, aucell){
  colnames(aucell) <- gsub("[...]", "", colnames(aucell))
  exp <- t(as.matrix(aucell))
  #Get rid of odd case of all zero values
  exp <- exp[rowSums(exp)!=0,]
  exp <- t(scale(t(exp), scale=T, center=T))
  
  gam.coef <- apply(exp,1,function(z){
    d <- data.frame(exp=z, pseudotime=seurat_object$Pseudotime,
                    batch = seurat_object$Mix)
    tmp <- mgcv::gam(exp ~ pseudotime + batch, data=d)
    p <- coef(tmp)[2] # where the pseudotime coefficient is stored
    p
    })
  gam.coef <- as.data.frame(gam.coef)
  colnames(gam.coef) <- "Coefficient"
  return(gam.coef)
}
```

```{r}
a549_tgfb1_coef <- linearGAM(a549_tgfb1, a549_tgfb1_aucell)
a549_egf_coef <- linearGAM(a549_egf, a549_egf_aucell)
a549_tnf_coef <- linearGAM(a549_tnf, a549_tnf_aucell)

du145_tgfb1_coef <- linearGAM(du145_tgfb1, du145_tgfb1_aucell)
du145_egf_coef <- linearGAM(du145_egf, du145_egf_aucell)
du145_tnf_coef <- linearGAM(du145_tnf, du145_tnf_aucell)

mcf7_tgfb1_coef <- linearGAM(mcf7_tgfb1, mcf7_tgfb1_aucell)
mcf7_egf_coef <- linearGAM(mcf7_egf, mcf7_egf_aucell)
mcf7_tnf_coef <- linearGAM(mcf7_tnf, mcf7_tnf_aucell)

ovca420_tgfb1_coef <- linearGAM(ovca420_tgfb1, ovca420_tgfb1_aucell)
ovca420_egf_coef <- linearGAM(ovca420_egf, ovca420_egf_aucell)
ovca420_tnf_coef <- linearGAM(ovca420_tnf, ovca420_tnf_aucell)
```

```{r}
conserved_genes <- as.character(filter(deg_list_counts, Count >= 6)$Gene)
```


```{r}
coef_mat <- data.frame(A549_TGFB1 = a549_tgfb1_coef[conserved_genes,"Coefficient"],
                       A549_EGF = a549_egf_coef[conserved_genes, "Coefficient"],
                       A549_TNF = a549_tnf_coef[conserved_genes, "Coefficient"],
                       DU145_TGFB1 = du145_tgfb1_coef[conserved_genes,"Coefficient"],
                       DU145_EGF = du145_egf_coef[conserved_genes, "Coefficient"],
                       DU145_TNF = du145_tnf_coef[conserved_genes, "Coefficient"],
                       MCF7_TGFB1 = mcf7_tgfb1_coef[conserved_genes,"Coefficient"],
                       MCF7_EGF = mcf7_egf_coef[conserved_genes, "Coefficient"],
                       MCF7_TNF = mcf7_tnf_coef[conserved_genes, "Coefficient"],
                       OVCA420_TGFB1 = ovca420_tgfb1_coef[conserved_genes,"Coefficient"],
                       OVCA420_EGF = ovca420_egf_coef[conserved_genes, "Coefficient"],
                       OVCA420_TNF = ovca420_tnf_coef[conserved_genes, "Coefficient"])
coef_mat <- as.matrix(coef_mat)
rownames(coef_mat) <- conserved_genes
coef_mat[is.na(coef_mat)] <- 0

coef_mat[coef_mat>1] <- 1
coef_mat[coef_mat < (-1)] <- (-1)

#Remove any regulons that aren't significantly changed
temp_deg_list <- deg_list
rownames(temp_deg_list) <- temp_deg_list$Gene
temp_deg_list$Gene <- NULL
temp_deg_list <- as.matrix(temp_deg_list)
temp_deg_list <- temp_deg_list[conserved_genes,]

coef_mat <- coef_mat * temp_deg_list
```

```{r}
coef_heatmap <- pheatmap(coef_mat,
                         color = colorRampPalette(rev(RColorBrewer::brewer.pal(7, "RdBu")))(100),
                         border_color = "black",
                         cluster_cols=F,
                         cluster_rows=T,
                         clustering_method = "ward.D2",
                         show_rownames=T,
                         show_colnames=F,
                         legend=F,
                         treeheight_row = 0,
                         fontsize=8,
                         filename="../figure_panels/figure_3/Pseudotime_regulon_coefficient.png",
                         width=2.55, height=3.1)
```


# Fig 3d
## ATAC differential accessibility
```{r}
atac_mat <- read.csv("../output/ATAC_differential_accessibility.csv",
                     row.names=1)
atac_mat <- as.matrix(atac_mat)
```

```{r}
motif.heatmap <- pheatmap(atac_mat,
                    color=viridis::viridis(100),
                    scale="none",
                    cluster_rows=T,
                    clustering_method = "ward.D2",
                    cutree_rows=6,
                    treeheight_row = 0,
                    cluster_cols=F,
                    show_rownames=F,
                    show_colnames=F,
                    legend=F,
                    filename = "../figure_panels/figure_3/ATAC_diff_accessibility.png",
                    width=2, height=3.95)
```

And small little script to find out how far in each of the 6 clusters the various
genes of interest are

```{r}
clusters <- as.data.frame(cutree(motif.heatmap$tree_row, k=6))
colnames(clusters) <- "Cluster"
clusters$Gene <- rownames(clusters)
```

Order of the clusters
```{r}
unique(clusters$Cluster[motif.heatmap$tree_row$order])
```

```{r}
table(clusters$Cluster)
```


```{r}
clusters <- clusters[motif.heatmap$tree_row$order,]
which(filter(clusters, Cluster==6)$Gene == "TP63")
```

```{r}
which(filter(clusters, Cluster==2)$Gene == "STAT5B")
```

```{r}
which(filter(clusters, Cluster==1)$Gene == "RELB")
```

```{r}
which(filter(clusters, Cluster==3)$Gene == "BRCA1")
```

```{r}
which(filter(clusters, Cluster==4)$Gene == "FOXA2")
```

```{r}
which(filter(clusters, Cluster==5)$Gene == "CEBPD")
```


# Fig 3e
## Regulon score
```{r}
PRGn <- rev(RColorBrewer::brewer.pal(11, "PRGn"))
color_ramp <- c(PRGn[1], PRGn[2], PRGn[3], PRGn[4], 'grey50', PRGn[7], PRGn[8], PRGn[9])
```

```{r}
plotAUC <- function(seurat_object, aucell, gene, sample){
  colnames(aucell) <- gsub("[...]", "", colnames(aucell))
  data <- data.frame(Pseudotime = seurat_object$Pseudotime,
                     Activity = aucell[,gene],
                     Sample = seurat_object$Sample)
  data_plot <- ggplot(data, aes(x=Pseudotime, y=Activity)) +
    geom_point(size=0.1, shape=16, aes(color=Sample)) +
    geom_smooth(color="black") +
    scale_color_manual(values=color_ramp) +
    scale_x_continuous(expand=c(0,0), breaks=c(0, 0.5, 1)) +
    scale_y_continuous(expand=c(0,0)) +
    ylab("Regulon AUC") +
    theme_classic() +
    theme(legend.position="none",
          axis.text = element_text(size=8, color="black"),
          axis.title = element_blank(),
          axis.line = element_line(size=0.25),
          axis.ticks = element_line(size=0.25))
  ggsave(data_plot, filename=paste0("../figure_panels/figure_3/", "RegulonActivity_",
                                    sample, "_", gene, ".png"),
         width=1.2, height=1.4)
  data_plot
}
```

```{r}
# Top ATAC cluster (cluster 6)
p1 <- plotAUC(ovca420_tgfb1, ovca420_tgfb1_aucell, "TP63", "OVCA420_TGFB1")

# Second ATAC cluster (Cluster 2)
p2 <- plotAUC(ovca420_tgfb1, ovca420_tgfb1_aucell, "STAT5B", "OVCA420_TGFB1") # Not the best fit

# Third ATAC cluster (Cluster 1)
p3 <- plotAUC(ovca420_tgfb1, ovca420_tgfb1_aucell, "RELB", "OVCA420_TGFB1")

# Fouth ATAC cluster (Cluster 3)
p4 <- plotAUC(ovca420_tgfb1, ovca420_tgfb1_aucell, "BRCA1", "OVCA420_TGFB1") #CTCF is another good one

# Fifth ATAC cluster (Cluster 4)
p5 <- plotAUC(ovca420_tgfb1, ovca420_tgfb1_aucell, "FOXA2", "OVCA420_TGFB1")

# Sixth ATAC cluster (Cluster 5)
p6 <- plotAUC(ovca420_tgfb1, ovca420_tgfb1_aucell, "CEBPD", "OVCA420_TGFB1") #EGR2 maybe
```

```{r}
grid_plot <- plot_grid(p1, p2, p3, p4, p5, p6,
                       ncol = 3)
save_plot(grid_plot, base_width=7.75, base_height=7,
          filename = "../figs/OVCA420_TGFB1/SCENIC/Regulon_Grid.png",
          dpi=600)
save_plot(grid_plot, base_width=9, base_height=7.5,
          filename = "../figs/OVCA420_TGFB1/SCENIC/Regulon_Grid.pdf")
```

# Session Info
```{r}
sessionInfo()
```

