---
title: "Figure 4"
output: html_notebook
---

# Dependencies
```{r}
library(Seurat)
library(mgcv)
library(ggplot2)
library(dplyr)
library(viridis)
library(pheatmap)
library(fgsea)
library(ggsci)
library(stringr)
library(RcisTarget)
```

# Load the data
Timecourse data
```{r}
a549_egf <- readRDS("../data/A549_EGF.rds")
a549_tnf <- readRDS("../data/A549_TNF.rds")
a549_tgfb1 <- readRDS("../data/A549_TGFB1.rds")

du145_egf <- readRDS("../data/DU145_EGF.rds")
du145_tnf <- readRDS("../data/DU145_TNF.rds")
du145_tgfb1 <- readRDS("../data/DU145_TGFB1.rds")

mcf7_egf <- readRDS("../data/MCF7_EGF.rds")
mcf7_tnf <- readRDS("../data/MCF7_TNF.rds")
mcf7_tgfb1 <- readRDS("../data/MCF7_TGFB1.rds")

ovca420_egf <- readRDS("../data/OVCA420_EGF.rds")
ovca420_tnf <- readRDS("../data/OVCA420_TNF.rds")
ovca420_tgfb1 <- readRDS("../data/OVCA420_TGFB1.rds")
```

Kinase inhibitor data
```{r}
a549_untreated_kinase <- readRDS("../data/A549_Kinase_Untreated.rds")
a549_tgfb1_kinase <- readRDS("../data/A549_Kinase_TGFB1.rds")
a549_egf_kinase <- readRDS("../data/A549_Kinase_EGF.rds")
a549_tnf_kinase <- readRDS("../data/A549_Kinase_TNF.rds")

du145_untreated_kinase <- readRDS("../data/DU145_Kinase_Untreated.rds")
du145_tgfb1_kinase <- readRDS("../data/DU145_Kinase_TGFB1.rds")
du145_egf_kinase <- readRDS("../data/DU145_Kinase_EGF.rds")
du145_tnf_kinase <- readRDS("../data/DU145_Kinase_TNF.rds")

mcf7_untreated_kinase <- readRDS("../data/MCF7_Kinase_Untreated.rds")
mcf7_tgfb1_kinase <- readRDS("../data/MCF7_Kinase_TGFB1.rds")
mcf7_egf_kinase <- readRDS("../data/MCF7_Kinase_EGF.rds")
mcf7_tnf_kinase <- readRDS("../data/MCF7_Kinase_TNF.rds")

ovca420_untreated_kinase <- readRDS("../data/OVCA420_Kinase_Untreated.rds")
ovca420_tgfb1_kinase <- readRDS("../data/OVCA420_Kinase_TGFB1.rds")
ovca420_egf_kinase <- readRDS("../data/OVCA420_Kinase_EGF.rds")
ovca420_tnf_kinase <- readRDS("../data/OVCA420_Kinase_TNF.rds")
```

# Figure 4a
## Cytokine score
```{r}
cytokines <- fgsea::gmtPathways("~/Data/GeneLists/KEGG_chemokine_chemokine_receptor_interaction.gmt")
cytokines <- cytokines[[1]]
```

First identify which of the cytokines are differentially expressed

```{r}
# For the actual test, we'll assume linear relationship because we're trying to focus on changes
# that are actually mesenchymal- or epithelial-associated, rather than transient changes
testGAM <- function(seurat_object, genes){
  genes <- genes[which(genes %in% rownames(seurat_object))]
  exp <- as.matrix(seurat_object[["RNA"]]@data[genes,])
  
  gam.pval <- apply(exp,1,function(z){
    d <- data.frame(exp=z, pseudotime=seurat_object$Pseudotime,
                    batch=seurat_object$Mix)
    tmp <- mgcv::gam(exp ~ pseudotime + batch, 
                     data=d)
    p <- summary(tmp)[[4]][2] # where the p-value is stored
    p
    })
  gam.pval <- as.data.frame(gam.pval)
  colnames(gam.pval) <- "pval"
  gam.pval$Gene <- rownames(gam.pval)
  gam.pval$fdr <- p.adjust(gam.pval$pval, method='fdr')
  return(gam.pval)
}
```

Test for sig cytokines
```{r}
a549_tgfb1_cyto_pval <- testGAM(a549_tgfb1, cytokines)
a549_egf_cyto_pval <- testGAM(a549_egf, cytokines)
a549_tnf_cyto_pval <- testGAM(a549_tnf, cytokines)

du145_tgfb1_cyto_pval <- testGAM(du145_tgfb1, cytokines)
du145_egf_cyto_pval <- testGAM(du145_egf, cytokines)
du145_tnf_cyto_pval <- testGAM(du145_tnf, cytokines)

mcf7_tgfb1_cyto_pval <- testGAM(mcf7_tgfb1, cytokines)
mcf7_egf_cyto_pval <- testGAM(mcf7_egf, cytokines)
mcf7_tnf_cyto_pval <- testGAM(mcf7_tnf, cytokines)

ovca420_tgfb1_cyto_pval <- testGAM(ovca420_tgfb1, cytokines)
ovca420_egf_cyto_pval <- testGAM(ovca420_egf, cytokines)
ovca420_tnf_cyto_pval <- testGAM(ovca420_tnf, cytokines)
```

Get sig genes
```{r}
a549_tgfb1_sig <- rownames(a549_tgfb1_cyto_pval)[which(a549_tgfb1_cyto_pval$fdr <= 0.05)]
a549_egf_sig <- rownames(a549_egf_cyto_pval)[which(a549_egf_cyto_pval$fdr <= 0.05)]
a549_tnf_sig <- rownames(a549_tnf_cyto_pval)[which(a549_tnf_cyto_pval$fdr <= 0.05)]

du145_tgfb1_sig <- rownames(du145_tgfb1_cyto_pval)[which(du145_tgfb1_cyto_pval$fdr <= 0.05)]
du145_egf_sig <- rownames(du145_egf_cyto_pval)[which(du145_egf_cyto_pval$fdr <= 0.05)]
du145_tnf_sig <- rownames(du145_tnf_cyto_pval)[which(du145_tnf_cyto_pval$fdr <= 0.05)]

mcf7_tgfb1_sig <- rownames(mcf7_tgfb1_cyto_pval)[which(mcf7_tgfb1_cyto_pval$fdr <= 0.05)]
mcf7_egf_sig <- rownames(mcf7_egf_cyto_pval)[which(mcf7_egf_cyto_pval$fdr <= 0.05)]
mcf7_tnf_sig <- rownames(mcf7_tnf_cyto_pval)[which(mcf7_tnf_cyto_pval$fdr <= 0.05)]

ovca420_tgfb1_sig <- rownames(ovca420_tgfb1_cyto_pval)[which(ovca420_tgfb1_cyto_pval$fdr <= 0.05)]
ovca420_egf_sig <- rownames(ovca420_egf_cyto_pval)[which(ovca420_egf_cyto_pval$fdr <= 0.05)]
ovca420_tnf_sig <- rownames(ovca420_tnf_cyto_pval)[which(ovca420_tnf_cyto_pval$fdr <= 0.05)]

sig_cytokines <- unique(c(a549_tgfb1_sig, a549_egf_sig, a549_tnf_sig,
                          du145_tgfb1_sig, du145_egf_sig, du145_tnf_sig,
                          mcf7_tgfb1_sig, mcf7_egf_sig, mcf7_tnf_sig,
                          ovca420_tgfb1_sig, ovca420_egf_sig, ovca420_tnf_sig))
```

```{r}
scoreGeneSet <- function(seurat_object){
  seurat_object <- AddModuleScore(seurat_object, 
                                  features = list(sig_cytokines),
                                  name="Cytokine_Score")
  return(seurat_object)
}
```

```{r}
a549_tgfb1 <- scoreGeneSet(a549_tgfb1)
a549_egf <- scoreGeneSet(a549_egf)
a549_tnf <- scoreGeneSet(a549_tnf)

du145_tgfb1 <- scoreGeneSet(du145_tgfb1)
du145_egf <- scoreGeneSet(du145_egf)
du145_tnf <- scoreGeneSet(du145_tnf)

mcf7_tgfb1 <- scoreGeneSet(mcf7_tgfb1)
mcf7_egf <- scoreGeneSet(mcf7_egf)
mcf7_tnf <- scoreGeneSet(mcf7_tnf)

ovca420_tgfb1 <- scoreGeneSet(ovca420_tgfb1)
ovca420_egf <- scoreGeneSet(ovca420_egf)
ovca420_tnf <- scoreGeneSet(ovca420_tnf)
```

```{r}
plotGeneSetScore <- function(geneset){
  data <- bind_rows(a549_tgfb1@meta.data, a549_egf@meta.data, a549_tnf@meta.data,
                    du145_tgfb1@meta.data, du145_egf@meta.data, du145_tnf@meta.data,
                    mcf7_tgfb1@meta.data, mcf7_egf@meta.data, mcf7_tnf@meta.data,
                    ovca420_tgfb1@meta.data, ovca420_egf@meta.data, ovca420_tnf@meta.data)
  #data$Group <- paste0(data$CellLine, "_", data$Treatment)
  shuffle <- sample(1:nrow(data), replace=F)
  data <- data[shuffle,]
  data$Treatment <- factor(data$Treatment, levels=c("TGFB1", "EGF", "TNF"))
  
  data_plot <- ggplot(data, aes_string(x="Pseudotime", y=paste0(geneset, "1"))) +
    #geom_point(size=0.1, alpha=0.25, color="grey20") +
    geom_smooth(size=0.5, aes(color=Treatment)) +
    scale_x_continuous(expand=c(0,0), breaks = c(0,0.5,1)) +
    scale_y_continuous(expand=c(0,0)) +
    scale_color_nejm() +
    ylab("EMT Score") +
    facet_wrap(~CellLine, scales="free", ncol = 2) +
    theme_classic() +
    theme(strip.background = element_blank(),
          strip.text = element_blank(), #I want to put the labels in myself
          legend.position="none",
          axis.title=element_blank(),
          axis.text = element_text(size=6, color="black"),
          axis.line = element_line(size=0.15),
          axis.ticks = element_line(size=0.15))
  data_plot
}
```

```{r}
cytokine_plot <- plotGeneSetScore("Cytokine_Score")
cytokine_plot
```

```{r}
ggsave(cytokine_plot, filename = "../figure_panels/figure_4/Cytokine_Score.png",
       units="cm", width=4.5, height=4.85, dpi=600)
```

# Figure 4b
## Cytokine heatmap
Heatmap of linear pseudotime coefficient

```{r}
linearGAM <- function(seurat_object, genes){
  genes <- genes[which(genes %in% rownames(seurat_object))]
  exp <- as.matrix(seurat_object[["RNA"]]@data[genes,])
  
  gam.coef <- apply(exp,1,function(z){
    d <- data.frame(exp=z, pseudotime=seurat_object$Pseudotime,
                    batch = seurat_object$Mix)
    tmp <- mgcv::gam(exp ~ pseudotime + batch, data=d)
    p <- coef(tmp)[2] # where the pseudotime coefficient is stored
    p
    })
  gam.coef <- as.data.frame(gam.coef)
  colnames(gam.coef) <- "Coefficient"
  return(gam.coef)
}
```

```{r}
a549_tgfb1_cyto_coef <- linearGAM(a549_tgfb1, cytokines)
a549_egf_cyto_coef <- linearGAM(a549_egf, cytokines)
a549_tnf_cyto_coef <- linearGAM(a549_tnf, cytokines)

du145_tgfb1_cyto_coef <- linearGAM(du145_tgfb1, cytokines)
du145_egf_cyto_coef <- linearGAM(du145_egf, cytokines)
du145_tnf_cyto_coef <- linearGAM(du145_tnf, cytokines)

mcf7_tgfb1_cyto_coef <- linearGAM(mcf7_tgfb1, cytokines)
mcf7_egf_cyto_coef <- linearGAM(mcf7_egf, cytokines)
mcf7_tnf_cyto_coef <- linearGAM(mcf7_tnf, cytokines)

ovca420_tgfb1_cyto_coef <- linearGAM(ovca420_tgfb1, cytokines)
ovca420_egf_cyto_coef <- linearGAM(ovca420_egf, cytokines)
ovca420_tnf_cyto_coef <- linearGAM(ovca420_tnf, cytokines)
```

```{r}
coef_mat <- data.frame(A549_TGFB1 = a549_tgfb1_cyto_coef[sig_cytokines,"Coefficient"],
                       A549_EGF = a549_egf_cyto_coef[sig_cytokines, "Coefficient"],
                       A549_TNF = a549_tnf_cyto_coef[sig_cytokines, "Coefficient"],
                       DU145_TGFB1 = du145_tgfb1_cyto_coef[sig_cytokines,"Coefficient"],
                       DU145_EGF = du145_egf_cyto_coef[sig_cytokines, "Coefficient"],
                       DU145_TNF = du145_tnf_cyto_coef[sig_cytokines, "Coefficient"],
                       MCF7_TGFB1 = mcf7_tgfb1_cyto_coef[sig_cytokines,"Coefficient"],
                       MCF7_EGF = mcf7_egf_cyto_coef[sig_cytokines, "Coefficient"],
                       MCF7_TNF = mcf7_tnf_cyto_coef[sig_cytokines, "Coefficient"],
                       OVCA420_TGFB1 = ovca420_tgfb1_cyto_coef[sig_cytokines,"Coefficient"],
                       OVCA420_EGF = ovca420_egf_cyto_coef[sig_cytokines, "Coefficient"],
                       OVCA420_TNF = ovca420_tnf_cyto_coef[sig_cytokines, "Coefficient"])
coef_mat <- as.matrix(coef_mat)
rownames(coef_mat) <- sig_cytokines
coef_mat[is.na(coef_mat)] <- 0

coef_mat[coef_mat>0.25] <- 0.25
coef_mat[coef_mat < (-0.25)] <- (-0.25)
```

```{r}
coef_heatmap <- pheatmap(t(coef_mat),
                         color = colorRampPalette(rev(RColorBrewer::brewer.pal(7, "RdBu")))(100),
                         border_color = "black",
                         cluster_cols=T,
                         cluster_rows=F,
                         treeheight_col = 0,
                         clustering_method = "ward.D2",
                         show_rownames=F,
                         show_colnames=T,
                         legend=F,
                         fontsize=7,
                         treeheight_row = 0,
                         filename="../figure_panels/figure_4//Cytokine_pseudotime_coefficient.png",
                         width=11, height=2.755)
```

# Figure 4c
Schematic made by hand

# Figure 4d
## Cell counts for each condition
Have to set up a count table. a549_untreated_kinase has every drug represented, so we'll use that to start
```{r}
drug_list <- unique(a549_untreated_kinase$Drug)
drug_counts <- data.frame(a549_untreated = as.numeric(table(a549_untreated_kinase$Drug)[drug_list]),
                          a549_tgfb1 = as.numeric(table(a549_tgfb1_kinase$Drug)[drug_list]),
                          a549_egf = as.numeric(table(a549_egf_kinase$Drug)[drug_list]),
                          a549_tnf = as.numeric(table(a549_tnf_kinase$Drug)[drug_list]),
                          du145_untreated = as.numeric(table(du145_untreated_kinase$Drug)[drug_list]),
                          du145_tgfb1 = as.numeric(table(du145_tgfb1_kinase$Drug)[drug_list]),
                          du145_egf = as.numeric(table(du145_egf_kinase$Drug)[drug_list]),
                          du145_tnf = as.numeric(table(du145_tnf_kinase$Drug)[drug_list]),
                          mcf7_untreated = as.numeric(table(mcf7_untreated_kinase$Drug)[drug_list]),
                          mcf7_tgfb1 = as.numeric(table(mcf7_tgfb1_kinase$Drug)[drug_list]),
                          mcf7_egf = as.numeric(table(mcf7_egf_kinase$Drug)[drug_list]),
                          mcf7_tnf = as.numeric(table(mcf7_tnf_kinase$Drug)[drug_list]),
                          ovca420_untreated = as.numeric(table(ovca420_untreated_kinase$Drug)[drug_list]),
                          ovca420_tgfb1 = as.numeric(table(ovca420_tgfb1_kinase$Drug)[drug_list]),
                          ovca420_egf = as.numeric(table(ovca420_egf_kinase$Drug)[drug_list]),
                          ovca420_tnf = as.numeric(table(ovca420_tnf_kinase$Drug)[drug_list]))
rownames(drug_counts) <- drug_list
rownames(drug_counts)[which(rownames(drug_counts) == "Uninhibited_Untreated")] <- "No Inhibitor"
rownames(drug_counts)[which(rownames(drug_counts) == "Phthalazinone Pyrazole")] <- "P. Pyrazole"
drug_counts <- as.matrix(drug_counts)
drug_counts[is.na(drug_counts)] <- 0
```

```{r}
max(drug_counts)
```

```{r}
#Make a color key that caps out at 200
color_map <- c(viridis::magma(200), rep(viridis::magma(200)[200], (968-200)))

sample_count_heatmap <- pheatmap(t(drug_counts),
                                 color=color_map,
                                 cluster_rows=F,
                                 cluster_cols=T,
                                 treeheight_col = 0,
                                 show_rownames=F,
                                 show_colnames=T,
                                 legend=F,
                                 display_numbers=T,
                                 border_color="black",
                                 number_form="%.0f",
                                 number_color="grey50",
                                 file="../figure_panels/figure_4/kinase_screen_counts.png",
                                 width=5.5, height=4.25)
```


# Figure 4e
## DEG counts
```{r}
getCounts <- function(sample){
  coef <- read.csv(paste0("../output/kinase_differential_expression/", sample, "_coefficients.csv"),
                 row.names=1)
  pval <- read.csv(paste0("../output/kinase_differential_expression/", sample, "_padj.csv"),
                 row.names=1)
  
  pval <- as.matrix(pval)
  pval <- -log10(pval)
  pval[which(pval < (-log10(0.05)))] <- 0
  pval[which(pval > (-log10(0.05)))] <- 1
  pval <- pval[,-1] #remove intercept
  
  # Remove interaction hits
  interaction_coef_index <- grep("Treatment", colnames(pval))
  pval <- pval[,-interaction_coef_index]
  
  # Only include genes with sig hits somewhere
  pval <- pval[which(rowSums(pval) != 0),]
  
  #rename columns
  colnames(pval) <- str_replace(colnames(pval), ".*Drug", "")
  colnames(pval) <- str_replace(colnames(pval), "\\.", " ")
  
  sums <- colSums(pval)
  return(sums)
}
```

So technically we only have one untreated population per cell line, but because we have three different conditions for each cell line, we've also used three different sets of the top 5000 variable genes for differential expression testing, and there's also the possibility of things shifting because 
```{r}
a549_untreated_1 <- getCounts("A549_TGFB1")
a549_untreated_2 <- getCounts("A549_EGF")
a549_untreated_3 <- getCounts("A549_TNF")

du145_untreated_1 <- getCounts("DU145_TGFB1")
du145_untreated_2 <- getCounts("DU145_EGF")
du145_untreated_3 <- getCounts("DU145_TNF")

mcf7_untreated_1 <- getCounts("MCF7_TGFB1")
mcf7_untreated_2 <- getCounts("MCF7_EGF")
mcf7_untreated_3 <- getCounts("MCF7_TNF")

ovca420_untreated_1 <- getCounts("OVCA420_TGFB1")
ovca420_untreated_2 <- getCounts("OVCA420_EGF")
ovca420_untreated_3 <- getCounts("OVCA420_TNF")
```

They are incredibly well correlated
```{r}
plot(a549_untreated_1, a549_untreated_2)
```

I think it's fair to take the average of the three models as a reasonable count of the number of genes affected

```{r}
untreated_counts <- data.frame(A549 = (a549_untreated_1 + a549_untreated_2 + a549_untreated_3) / 3,
                               DU145 = (du145_untreated_1 + du145_untreated_2 + du145_untreated_3) / 3,
                               MCF7 = (mcf7_untreated_1 + mcf7_untreated_2 + mcf7_untreated_3) / 3,
                               OVCA420 = (ovca420_untreated_1 + ovca420_untreated_2 + ovca420_untreated_3) / 3)
rownames(untreated_counts) <- names(a549_untreated_1)

rownames(untreated_counts)[which(rownames(untreated_counts) == "Phthalazinone Pyrazole")] <- "P. Pyrazole"
```

```{r}
untreated_counts_cap <- untreated_counts
untreated_counts_cap[untreated_counts_cap > 500] <- 500

untreated_heatmap <- pheatmap(untreated_counts_cap,
                              color = viridis(100),
                              show_rownames=T,
                              show_colnames=F,
                              border_color="black",
                              cluster_rows=T,
                              cluster_cols=F,
                              treeheight_row=0,
                              treeheight_col=0,
                              fontsize=11,
                              legend=F,
                              clustering_method="ward.D2",
                              filename = "../figure_panels/figure_4/heatmap_inhibitor_untreated_counts.png",
                              width=2.1, height=4.4)
```

# Figure 4f
## Pseudotime values for inhibited samples

```{r}
a549_untreated_kinase_psuper <- readRDS("../output/pseudotime/a549_untreated_kinase_psuper.rds")
a549_tgfb_kinase_psuper <- readRDS("../output/pseudotime/a549_tgfb1_kinase_psuper.rds")
a549_egf_kinase_psuper <- readRDS("../output/pseudotime/a549_egf_kinase_psuper.rds")
a549_tnf_kinase_psuper <- readRDS("../output/pseudotime/a549_tnf_kinase_psuper.rds")

du145_untreated_kinase_psuper <- readRDS("../output/pseudotime/du145_untreated_kinase_psuper.rds")
du145_tgfb_kinase_psuper <- readRDS("../output/pseudotime/du145_tgfb1_kinase_psuper.rds")
du145_egf_kinase_psuper <- readRDS("../output/pseudotime/du145_egf_kinase_psuper.rds")
du145_tnf_kinase_psuper <- readRDS("../output/pseudotime/du145_tnf_kinase_psuper.rds")

mcf7_untreated_kinase_psuper <- readRDS("../output/pseudotime/mcf7_untreated_kinase_psuper.rds")
mcf7tgfb_kinase_psuper <- readRDS("../output/pseudotime/mcf7_tgfb1_kinase_psuper.rds")
mcf7_egf_kinase_psuper <- readRDS("../output/pseudotime/mcf7_egf_kinase_psuper.rds")
mcf7_tnf_kinase_psuper <- readRDS("../output/pseudotime/mcf7_tnf_kinase_psuper.rds")

ovca420_untreated_kinase_psuper <- readRDS("../output/pseudotime/ovca420_untreated_kinase_psuper.rds")
ovca420_tgfb_kinase_psuper <- readRDS("../output/pseudotime/ovca420_tgfb1_kinase_psuper.rds")
ovca420_egf_kinase_psuper <- readRDS("../output/pseudotime/ovca420_egf_kinase_psuper.rds")
ovca420_tnf_kinase_psuper <- readRDS("../output/pseudotime/ovca420_tnf_kinase_psuper.rds")
```


I want to rescale this so that the pseudotime values 0-1 correspond to the mean untreated and EMT-induced values, respectively
```{r}
summarizePseudo <- function(psuper, cell_line, treatment){
  summarized_data <- psuper %>%
    group_by(label_input) %>%
    summarize(pseudotime = mean(psuper))
  summarized_data$CellLine <- cell_line
  summarized_data$Treatment <- treatment
  
  #rescale
  untreated_mean <- summarized_data$pseudotime[which(summarized_data$label_input == "Uninhibited_Untreated")]
  summarized_data$pseudotime <- (summarized_data$pseudotime - untreated_mean)
  treated_mean <- summarized_data$pseudotime[which(summarized_data$label_input == paste0("Uninhibited_", treatment))]
  summarized_data$pseudotime <- summarized_data$pseudotime / treated_mean
  
  return(summarized_data)
}
```

```{r}
a549_tgfb1_summary <- summarizePseudo(a549_tgfb1_kinase_psuper, "A549", "TGFB1")
a549_egf_summary <- summarizePseudo(a549_egf_kinase_psuper, "A549", "EGF")
a549_tnf_summary <- summarizePseudo(a549_tnf_kinase_psuper, "A549", "TNF")

du145_tgfb1_summary <- summarizePseudo(du145_tgfb1_kinase_psuper, "DU145", "TGFB1")
du145_egf_summary <- summarizePseudo(du145_egf_kinase_psuper, "DU145", "EGF")
du145_tnf_summary <- summarizePseudo(du145_tnf_kinase_psuper, "DU145", "TNF")

mcf7_tgfb1_summary <- summarizePseudo(mcf7_tgfb1_kinase_psuper, "MCF7", "TGFB1")
mcf7_egf_summary <- summarizePseudo(mcf7_egf_kinase_psuper, "MCF7", "EGF")
mcf7_tnf_summary <- summarizePseudo(mcf7_tnf_kinase_psuper, "MCF7", "TNF")

ovca420_tgfb1_summary <- summarizePseudo(ovca420_tgfb1_kinase_psuper, "OVCA420", "TGFB1")
ovca420_egf_summary <- summarizePseudo(ovca420_egf_kinase_psuper, "OVCA420", "EGF")
ovca420_tnf_summary <- summarizePseudo(ovca420_tnf_kinase_psuper, "OVCA420", "TNF")
```

```{r}
pseudotime_summary <- bind_rows(a549_tgfb1_summary, a549_egf_summary, a549_tnf_summary,
                                du145_tgfb1_summary, du145_egf_summary, du145_tnf_summary,
                                mcf7_tgfb1_summary, mcf7_egf_summary, mcf7_tnf_summary,
                                ovca420_tgfb1_summary, ovca420_egf_summary, ovca420_tnf_summary)

pseudotime_summary$Condition <- paste0(pseudotime_summary$CellLine, "_", pseudotime_summary$Treatment)
```

Convert to a matrix for a heatmap
```{r}
pseudo_mat <- pseudotime_summary[,c("label_input", "Condition", "pseudotime")]
pseudo_mat <- as.data.frame(spread(pseudo_mat, Condition, pseudotime))
rownames(pseudo_mat) <- pseudo_mat$label_input
pseudo_mat$label_input <- NULL
pseudo_mat <- as.matrix(pseudo_mat)
#Note: We'll cap this heatmap at 0-1 for aesthetics. Most inhibitors fall within a reasonable range
#Though, some extend outside of it--just keep that in mind
pseudo_mat[pseudo_mat < 0] <- 0
pseudo_mat[pseudo_mat > 1] <- 1

pseudo_mat[is.na(pseudo_mat)] <- 1
pseudo_mat <- pseudo_mat[-which(rownames(pseudo_mat) %in% c("Uninhibited_EGF", "Uninhibited_TGFB1", 
                                              "Uninhibited_TNF","Uninhibited_Untreated")),]
rownames(pseudo_mat)[which(rownames(pseudo_mat) == "Phthalazinone Pyrazole")] <- "P. Pyrazole"

#reorder
pseudo_mat <- pseudo_mat[,c("A549_TGFB1", "A549_EGF", "A549_TNF",
                            "DU145_TGFB1", "DU145_EGF", "DU145_TNF",
                            "MCF7_TGFB1", "MCF7_EGF", "MCF7_TNF",
                            "OVCA420_TGFB1", "OVCA420_EGF", "OVCA420_TNF")]
```

```{r}
pseudo_heatmap <- pheatmap(pseudo_mat,
                           color = rev(viridis::magma(100)),
                           border_color="black",
                           cluster_rows=T,
                           cluster_cols=F,
                           clustering_method = "ward.D2",
                           treeheight_row = 0,
                           show_colnames=F,
                           show_rownames=T,
                           legend=F,
                           fontsize=9,
                           filename = "../figure_panels/figure_4/heatmap_inhibitor_pseudotime_values.png",
                           width=3.25, height=3.6)
```

# Figure 4g
## Boxplot of pseudotime values
```{r}
pseudoBoxplot <- function(data, cell_line, treatment){
  data$Treated <- factor(data$Treated, levels=c("Untreated", treatment))
  data$label_input <- as.character(data$label_input)
  data$label_input[data$label_input=="Phthalazinone Pyrazole"] <- "P. Pyrazole"
  data$label_input[data$label_input=="Uninhibited_Untreated"] <- "a1"
  data$label_input[data$label_input==paste0("Uninhibited_", treatment)] <- "a2"
  data$label_input <- factor(data$label_input)
  
  plot <- ggplot(data, aes(x=label_input, y=psuper)) + 
    geom_boxplot(aes(fill=Treated), outlier.size=0.1, size=0.25) +
    xlab("") + ylab("Pseudotime") +
    scale_x_discrete(labels=c("a1" = "Untreated", "a2" = treatment)) +
    scale_fill_manual(values=c("grey65", RColorBrewer::brewer.pal(4, "Dark2")[2]))+
    theme_bw() +
    theme(axis.text.x=element_text(angle=45, hjust=1, color="black", size=9),
          axis.text.y=element_text(size=9, colour="black"),
          axis.title=element_blank(),
          legend.position="none")
  ggsave(plot, filename = paste0("../figure_panels/figure_4/Pseudotime_", 
                                 cell_line, "_", treatment,  "_withUntreated.png"),
         dpi=600, width=6, height=3.25)
  plot
}
```

```{r}
pseudoBoxplot(a549_untreated_kinase_psuper, "A549", "TGFB1")
```


# Figure 4h
## UMAP of a couple inhibitors
```{r}
plotUMAP <- function(cell_line, treatment, drugs){
  drug_list <- c("Uninhibited_Untreated", paste0("Uninhibited_", treatment), drugs)
  seurat <- readRDS(paste0("../data/", cell_line, "_Kinase_", treatment, ".rds"))
  cells_keep <- colnames(seurat)[seurat$Drug %in% drug_list]
  seurat <- subset(seurat, cells=cells_keep)
  seurat <- FindVariableFeatures(seurat, selection.method = "vst",
                               nfeatures=2000)
  seurat <- ScaleData(seurat, features=VariableFeatures(seurat),
                    vars.to.regress=c("nCount_RNA", "percent.mito",
                                      "S.Score", "G2M.Score"))
  seurat <- RunPCA(seurat, verbose=F)
  seurat <- RunUMAP(seurat, dims=1:30)
  
  data <- seurat@meta.data
  data$UMAP1 <- Embeddings(seurat, "umap")[,1]
  data$UMAP2 <- Embeddings(seurat, "umap")[,2]
  data$Drug <- factor(data$Drug, levels=drug_list)
  
  umap <- ggplot(data, aes(x=UMAP1, y=UMAP2)) +
    geom_point(size=0.5, aes(color=Drug)) +
    scale_color_discrete(labels=c("Untreated", treatment, drugs)) +
    theme_void() +
    theme(legend.position="none")
  
  ggsave(umap, filename=paste0("../figure_panels/figure_4/UMAP_", 
                               cell_line, "_", treatment, "_", drugs, ".png"),
       width=4, height=2.75)
  umap
}
```

```{r}
plotUMAP("A549", "TGFB1", "Necrostatin-5")
plotUMAP("A549", "TGFB1", "LY364947")
```



# Figure 4i
## Expression heatmap
```{r}
getPseudobulk <- function(cell_line, treatment, inhibitor){
  p_values <- read.csv(paste0("../output/kinase_differential_expression/", cell_line, "_",
                             treatment, "_padj.csv"), row.names = 1)
  gene_hits <- rownames(p_values)[which(p_values[,2] < 0.05)]
  
  seurat <- readRDS(paste0("../data/", cell_line, "_Kinase_", treatment,
                           ".rds"))
  untreated_cells <- colnames(seurat)[seurat$ConditionBroad == "Untreated_Uninhibited"]
  treated_cells <- colnames(seurat)[seurat$ConditionBroad == paste0(treatment, "_Uninhibited")]
  inhibited_cells <- colnames(seurat)[seurat$Drug == inhibitor]
  
  untreated_avg <- rowMeans(as.matrix(seurat[["RNA"]]@data[gene_hits,untreated_cells]))
  treated_avg <- rowMeans(as.matrix(seurat[["RNA"]]@data[gene_hits,treated_cells]))
  inhibited_avg <- rowMeans(as.matrix(seurat[["RNA"]]@data[gene_hits,inhibited_cells]))
  
  exp_avg <- data.frame(Untreated = untreated_avg,
                        Treated = treated_avg,
                        Inhibited = inhibited_avg)
  colnames(exp_avg) <- c("Untreated", treatment, paste0(treatment, "_", inhibitor))
  rownames(exp_avg) <- gene_hits
  exp_avg <- as.matrix(exp_avg)
  
  exp_mat <- t(scale(t(exp_avg), scale=T, center=T))
  exp_heatmap <- pheatmap(t(exp_mat),
                        color = viridis(100),
                        cluster_rows=F,
                        cluster_cols=T,
                        cutree_cols = 4,
                        treeheight_col = 25,
                        clustering_method="ward.D2",
                        show_rownames=F,
                        show_colnames=F,
                        legend = F,
                        filename=paste0("../figure_panels/figure_4/pseudobulk_",
                                        cell_line, "_", treatment, "_", inhibitor, ".png"),
                        width=5, height=1.25)
  return(exp_heatmap)
}
```

```{r}
a549_tgfb1_necrostatin <- getPseudobulk("A549", "TGFB1", "Necrostatin-5")
```

# Figure 4j
## cisTarget on heatmap
```{r}
clusters <- as.data.frame(cutree(a549_tgfb1_necrostatin$tree_col, k=4))
colnames(clusters) <- "Cluster"
clusters$Gene <- rownames(clusters)
table(clusters$Cluster)
```

Order of the clustering:
```{r}
unique(clusters$Cluster[a549_tgfb1_necrostatin$tree_col$order])
```

```{r}
motifRankings <- importRankings("~/Data/GeneLists/cisTarget/hg19-tss-centered-10kb-10species.mc9nr.feather")
```

```{r}
motifAnnotations <- importAnnotations("~/Data/GeneLists/cisTarget/motifs-v9-nr.hgnc-m0.001-o0.0.tbl.txt")
```

```{r}
getTF <- function(genes){
  gene_list <- list(SigGenes=genes)
  motif_enrichment <- cisTarget(gene_list, 
         motifRankings,
         motifAnnot=motifAnnotations,
         nesThreshold=0)
  motif_scores <- data.frame(motif = motif_enrichment$motif,
                             NES = motif_enrichment$NES,
                             TF_highConf = motif_enrichment$TF_highConf,
                             TF_lowConf = motif_enrichment$TF_lowConf)
  
  return(motif_scores)
}
```

Compare cluster 1 (TGFB1 up, Nec-5 blocks) and Cluster 2 (TGFB1 up, Nec-5 doesn't affect)

```{r}
cluster1_genes <- filter(clusters, Cluster==1)$Gene
cluster2_genes <- filter(clusters, Cluster==2)$Gene
```

```{r}
cluster1_tf <- getTF(cluster1_genes)
cluster2_tf <- getTF(cluster2_genes)

cluster_1_2 <- left_join(cluster1_tf, cluster2_tf, by="motif")
cluster_1_2$NES_diff <- cluster_1_2$NES.x - cluster_1_2$NES.y
cluster_1_2 <- arrange(cluster_1_2, desc(NES_diff))
```

```{r}
head(arrange(cluster_1_2, desc(NES_diff)), n=25)
```

```{r}
head(arrange(cluster_1_2, NES_diff), n=25)
```

Compare cluster 3 (TGFB1 down, Nec-5 blocks) and Cluster 4 (TGFB1 down, Nec-5 doesn't affect)

```{r}
cluster3_genes <- filter(clusters, Cluster==3)$Gene
cluster4_genes <- filter(clusters, Cluster==4)$Gene
```

```{r}
cluster3_tf <- getTF(cluster3_genes)
cluster4_tf <- getTF(cluster4_genes)

cluster_3_4 <- left_join(cluster3_tf, cluster4_tf, by="motif")
cluster_3_4$NES_diff <- cluster_3_4$NES.x - cluster_3_4$NES.y
cluster_3_4 <- arrange(cluster_3_4, desc(NES_diff))
```

```{r}
head(arrange(cluster_3_4, desc(NES_diff)), n=25)
```

```{r}
head(arrange(cluster_3_4, NES_diff), n=25)
```

Plot NES_diff curves
```{r}
cluster_1_2 <- na.omit(cluster_1_2)
cluster_1_2$Rank <- 1:nrow(cluster_1_2)
cluster1_2_plot <- ggplot(cluster_1_2, aes(x=Rank, y=NES_diff)) +
  geom_point(shape=16, size=0.2) +
  xlab("Motif Rank") + ylab("NES Difference") +
  theme_classic() +
  theme(axis.text = element_text(size=12, color="black"),
        axis.title=element_blank(),
        axis.line = element_line(size=0.35),
        axis.ticks = element_line(size=0.35))
cluster1_2_plot
ggsave(cluster1_2_plot, filename = "../figure_panels/figure_4/NES_Difference_Cluster_TGFB1_upGenes.png",
       width=5, height=2)
```

```{r}
cluster_3_4 <- na.omit(cluster_3_4)
cluster_3_4$Rank <- 1:nrow(cluster_3_4)
cluster3_4_plot <- ggplot(cluster_3_4, aes(x=Rank, y=NES_diff)) +
  geom_point(shape=16, size=0.2) +
  xlab("Motif Rank") + ylab("NES Difference") +
  theme_classic() +
  theme(axis.text = element_text(size=12, color="black"),
        axis.title=element_blank(),
        axis.line = element_line(size=0.35),
        axis.ticks = element_line(size=0.35))
cluster3_4_plot
ggsave(cluster3_4_plot, filename = "../figure_panels/figure_4/NES_Difference_Cluster_TGFB1_downGenes.png",
       width=5, height=2)
```